test/AssetPool.test.js

```diff
@@ -975,4 +975,243 @@ describe("AssetPool", () => {
+        )
+      })
+    })
+
+    context("when governance hasn't upgraded to a new pool yet", () => {
+      it("should revert", async () => {
+        const amount = to1e18(100)
+        await assetPool.connect(underwriter1).deposit(amount)
+        await underwriterToken
+          .connect(underwriter1)
+          .approve(assetPool.address, amount)
+
+        const amountToUpgrade = to1e18(99)
+        await expect(
+          assetPool.connect(underwriter1).upgradeToNewAssetPool(amountToUpgrade)
+        ).to.be.revertedWith("New asset pool must be assigned")
+      })
+    })
+
+    context("when governance upgraded to a new pool", () => {
+      const amountToDeposit = to1e18(100)
+      const amountToUpgrade = to1e18(40)
+
+      context("when no collateral tokens were claimed by the pool", () => {
+        let tx
+
+        beforeEach(async () => {
+          await assetPool
+            .connect(coveragePool)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+
+          await assetPool.connect(underwriter1).deposit(amountToDeposit)
+          await underwriterToken
+            .connect(underwriter1)
+            .approve(assetPool.address, amountToDeposit)
+
+          tx = await assetPool
+            .connect(underwriter1)
+            .upgradeToNewAssetPool(amountToUpgrade)
+        })
+
+        it("should transfer collateral tokens to the new asset pool", async () => {
+          expect(
+            await collateralToken.balanceOf(newAssetPool.address)
+          ).to.equal(amountToUpgrade)
+        })
+
+        it("should transfer new underwriter tokens to the underwriter", async () => {
+          expect(
+            await newUnderwriterToken.balanceOf(underwriter1.address)
+          ).to.equal(amountToUpgrade)
+        })
+
+        it("should burn old underwriter tokens from the underwriter", async () => {
+          // 100 - 40 = 60
+          expect(
+            await underwriterToken.balanceOf(underwriter1.address)
+          ).to.equal(amountToDeposit.sub(amountToUpgrade))
+        })
+
+        it("should emit AssetPoolUpgraded event", async () => {
+          expect(tx)
+            .to.emit(assetPool, "AssetPoolUpgraded")
+            .withArgs(
+              underwriter1.address,
+              amountToUpgrade,
+              amountToUpgrade,
+              await lastBlockTime()
+            )
+        })
+      })
+      context("when there was a claim", () => {
+        const claim = to1e18(25)
+
+        beforeEach(async () => {
+          await assetPool
+            .connect(coveragePool)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+          await assetPool.connect(underwriter1).deposit(amountToDeposit)
+          await assetPool
+            .connect(coveragePool)
+            .claim(coveragePool.address, claim)
+          await underwriterToken
+            .connect(underwriter1)
+            .approve(assetPool.address, amountToDeposit)
```

Comment:
Shouldn't `amountToUpgrade` be approved instead?

Reply:
18fed51